Improved error handling when GDAL 2.x fails to open existing layer

snorfalorpagus opened the pull request:
This PR provides slightly more informative error messages when failing to open a dataset in read mode. Previous more detailed messages from GDAL could be hidden by a generic (and potentially incorrect) message from Fiona.

I came across this issue in the 2.3.x (svn-trunk) job on Travis, which was raising this error:

```
fiona.errors.FionaValueError: No dataset found at path '/vsimem/c513cc7b5ce848a6a3358d4c0854f9bd.json' using drivers: {'GeoJSON'}
```

This is now replaced by a more correct message:

```
fiona.errors.FionaValueError: `/vsimem/8c013c21ebeb42cabfb0bb93239bf4d4.json' not recognized as a supported file format. using drivers {'GeoJSON'}
```

I think the actual problem is something in GDAL (see upstream issue https://trac.osgeo.org/gdal/ticket/7198 ), but the message Fiona gives is more useful now.

sgillies commented on fiona/_shim2.pyx:
```diff
@@ -49,10 +51,12 @@ cdef void* gdal_open_vector(char* path_c, int mode, drivers, options):
-    flags = GDAL_OF_VECTOR | mode
     try:
         cogr_ds = GDALOpenEx(
             path_c, flags, <const char *const *>drvs, open_opts, NULL)
```
@snorfalorpagus let's wrap the `GDALOpenEx` call in `exc_wrap_pointer()` https://github.com/Toblerity/Fiona/blob/master/fiona/_err.pyx#L222

snorfalorpagus commented:
This PR has grown a bit...

I've refactored the code used to open files for writing. I also fixed a bug in the condition used to detect GPKG support in the tests. This error was skipping some tests that were actually failing. I don't think writing binary fields has ever worked on Python 2.7, but does not. I think is is ready for review.

sgillies commented:
I'll review tomorrow morning.

sgillies commented on fiona/_shim1.pyx:
```diff
@@ -50,7 +50,10 @@ cdef void* gdal_open_vector(const char *path_c, int mode, drivers, options):
-        cogr_ds = OGROpen(path_c, mode, NULL)
+        try:
+            cogr_ds = exc_wrap_pointer(OGROpen(path_c, mode, NULL))
+        except Exception as exc:
```
I think we should catch only `CPLE_BaseError` here so we don't accidentally hide any errors related to malformed paths or bogus modes.

sgillies commented on fiona/_shim2.pyx:
```diff
@@ -49,11 +53,11 @@ cdef void* gdal_open_vector(char* path_c, int mode, drivers, options):
-        return cogr_ds
+        return exc_wrap_pointer(GDALOpenEx(
+            path_c, flags, <const char *const *>drvs, open_opts, NULL))
+    except Exception as exc:
```
`CPLE_BaseError` here too.

sgillies commented on fiona/ogrext.pyx:
```diff
@@ -335,6 +335,10 @@ cdef class OGRFeatureBuilder:
                 OGR_F_SetFieldDateTime(
                     cogr_feature, i, 0, 0, 0, hh, mm, ss, 0)
+            elif isinstance(value, bytes) and schema_type == "bytes":
+                string_c = value
```
@snorfalorpagus I think this is a good change, if we try to set a not-binary field to bytes we'll get a more opaque OGR error.

sgillies commented on fiona/ogrext.pyx:
```diff
@@ -420,26 +420,30 @@ cdef class Session:
-            self.cogr_ds = gdal_open_vector(path_c, 0, drivers, kwargs)
+        message = None
+        try:
+            with cpl_errs:
```
Could we use `exc_wrap_pointer` instead of `cpl_errs`?

sgillies commented on fiona/ogrext.pyx:
```diff
@@ -420,26 +420,30 @@ cdef class Session:
-                "No dataset found at path '%s' using drivers: %s" % (
-                    collection.path,
-                    drivers or '*'))
+                message + " using drivers {}".format(drivers or "*"))
```
👍

snorfalorpagus commented:
I tried to apply these changes to `_listlayers` but am getting some odd behavior:

```
>>> _listlayers("abc.json")
Traceback (most recent call last):
  File "fiona\_shim.pyx", line 63, in fiona._shim.gdal_open_vector
    return exc_wrap_pointer(GDALOpenEx(
  File "fiona\_err.pyx", line 229, in fiona._err.exc_wrap_pointer
    raise exc
fiona._err.CPLE_OpenFailedError: abc.json: No such file or directory

During handling of the above exception, another exception occurred:

fiona.errors.DriverIOError: abc.json: No such file or directory
Exception ignored in: 'fiona._shim.gdal_open_vector'
fiona.errors.DriverIOError: abc.json: No such file or directory
```

Can you hold off on this until I've worked this issue out.

snorfalorpagus commented:
Every time I think I've fixed something, something else breaks!

sgillies commented:
Don't sweat it @snorfalorpagus ! No rush.

snorfalorpagus commented:
I think I'm going to split some of these commits out into smaller PRs to make it more manageable.
